### ISSUE TITLE
Multivariate Taylor series expansion

### ISSUE THREAD

**Issue by ianhbell** [autodiff/autodiff#200]
I see you have implemented Taylor series expansion around a point in 1D, very nice. What I would like is to be able to do the same thing, efficiently, in two dimensions.  How difficult do you think it would be to implement efficiently in ``autodiff``? More generally, it would be nice to be able to fill in the matrix of terms d^(n+m)/d(x_1)^nd(x_2)^m all the terms up to n and m in the most efficient manner possible for a set of two input arguments. 

Context: I'm doing higher order thermodynamic variable transformations; it gets very hairy and I'd like to see if I can offload some of the pain onto ``autodiff``.

**Comment by allanleal** [comment]
Hi Ian, before I answer your question, let me give you some background.

`autodiff::real` is designed to be efficient in computing single-order and higher-order derivatives **along a specified direction**; it doesn't support cross derivatives like `autodiff::dual`, because:

* these are not required along a specific direction, and
* we want to avoid cross derivatives because their number increases exponentially with the derivative order of interest.

For example, `autodiff::real4th` only contains 5 floating-point values (one for the value itself, the other four for the derivatives up to 4th order). In contrast, `autodiff::dual4th` contains 16 floating-point values (to cover both the value and all derivatives, including the cross-derivatives). If 10th order, `autodiff::real10th` has 11 `double` values, and `autodiff::dual10th` has 1024.

I think you'll need to use `autodiff::dual` for your particular purpose (actually some `autodiff::dualNth`). For your two-variable Taylor expansion, you will need **cross derivatives**, which `autodiff::dual` provides. I'm not sure at this point how efficient (in speed and resources) this can be achieved, but the essential ingredients are there. 

I could design a new number type in `autodiff` with a specialized algorithm to handle this use-case more efficiently, but some sort of sponsorship would be needed! ;)

**Comment by allanleal** [comment]
> Conceptually, do you think that there would be a way to make the matrix evaluation I describe more efficient with the new datatype you propose?

I think it is possible to design a new number type that would computes all these derivatives simultaneously, instead of performing them through several calls of *Z(x, y)*. Right now, you would need to do something like this (inspired from this [tutorial](https://autodiff.github.io/tutorials/#higher-order-cross-derivatives-of-a-scalar-function):

~~~cpp
auto [Z, Zx, Zxx, Zxxx, Zxxxx] = derivatives(f, wrt(x, x, x, x), at(x, y));
auto [Z, Zx, Zxx, Zxxx, Zxxxy] = derivatives(f, wrt(x, x, x, y), at(x, y));
auto [Z, Zx, Zxx, Zxxy, Zxxyy] = derivatives(f, wrt(x, x, y, y), at(x, y));
auto [Z, Zx, Zxy, Zxyy, Zxyyy] = derivatives(f, wrt(x, y, y, y), at(x, y));
auto [Z, Zy, Zyy, Zyyy, Zyyyy] = derivatives(f, wrt(y, y, y, y), at(x, y));
~~~

**Attention*: The code above is just an idea; it does not compile and there are repeated variable names there.

Note that several identical derivatives are computed many times (e.g., `Zx`, `Zxx`), and `Z` is re-evaluated every time. A dedicated number type would compute Z and all its derivatives above in a single run (for any order). This should be more complicated than the algorithm I came up with for `autodiff::real` (I still have to publish it), but I think it should be possible to develop.

**Comment by ianhbell** [comment]
Yes, this is precisely what I am talking about - the duplication of effort with the terms that appear many times.

**Comment by allanleal** [comment]
Indeed, there is no way around this at the moment. Only a dedicated number type with its own specialized algorithm will be able to compute this efficiently. It would be a type that sits between the extreme use cases of `real` and `dual`.

**Comment by ianhbell** [comment]
I think in my very particular use case, most of the work can be done upfront and cached, so efficiency is not so paramount, but it would be nice to be able to have a nice interface for this problem.

**Comment by ianhbell** [comment]
Yes, that is entirely reasonable. It is not a totally standard application.
